### ISSUE TITLE
| reducing a set fails when + reducing it succeeds

### ISSUE THREAD

**Issue by cassella** [chapel-lang/chapel#24055]
### Summary of Problem

```chapel
use Set;

var combinations : set(int);

forall i in 1..1000 with (| reduce combinations) {
  for j in 1..100 {
    combinations.add(i*j);
  }
}

var serialcombs : set(int);
for i in 1..1000 do
  for j in 1..100 do
    serialcombs.add(i*j);

writeln(combinations == serialcombs);
```

**Compile command:**
`chpl foo.chpl`

```
> $CHPL_HOME/modules/standard/Types.chpl:994: In function '_bor_id':
> $CHPL_HOME/modules/standard/Types.chpl:994: error: illegal cast from int(64) to set(int(64),false)
>   $CHPL_HOME/modules/internal/ChapelReduce.chpl:298: called as _bor_id(type t = set(int(64),false))
>   within internal functions (use --print-callstack-on-error to see)
```

(from start_test)

This is the first half of the test submitted in #24054, which succeeds, except with the `+ reduced` changed to `| reduce`.

This is surprising to me because Set defines its `|` and `|=` operators the same as  its `+` and `+=` operators.  (In fact, the latter simply call the former.)

**Associated Future Test(s):**

None yet, pending agreement that this is a problem.

### Configuration Information

```
chpl version 1.33.0
  built with LLVM version 14.0.0
(really, today's main)

gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0

Ubuntu clang version 14.0.0-1ubuntu1.1

CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: llvm
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: qthreads
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: bundled
CHPL_HWLOC: bundled
CHPL_RE2: none
CHPL_LLVM: system
CHPL_AUX_FILESYS: none
```

**Comment by bradcray** [comment]
I think that this shows that the `+` operator is much more sophisticated at computing its identity in our reduction code than the `|` operator is.  @vasslitvinov: Does that sound right to you?  Is it a matter of replacing the call to `_bor_id()` in the `BitwiseOrReduceScanOp` with something more like `chpl_sumType()` in the `SumReduceScanOp`?

**Comment by vasslitvinov** [comment]
That's right. `|`-reduction gets its identity value from `_bor_id(element type)`.  `_bor_id` does not work for sets and most other types.

Cf. `+`-reduction gets the identify from the default initializer of the element type.

So I consider this error legitimate, just giving a poor error message.

Cf. `&`-reduction relies on `_band_id()` that prints a nice error message in this case. `_bor_id` could be updated in the same manner. While there, we could update `_band_id` to use transmute(). Also, inline `_land_id` and `_lor_id` and remove the unused `_lxor_id` to simplify code.

**Comment by bradcray** [comment]
> So I consider this error legitimate, just giving a poor error message.

@vasslitvinov : would adding a `_bor_id()` overload to `Set.chpl` that returned the empty set make this work?

**Comment by vasslitvinov** [comment]
That overload would not be visible from the reduction code (I just checked). This is because the reduction code in ChapelReduce.chpl does not `use` the Set module.

**Comment by bradcray** [comment]
I see…   :(  Is the situation better if we make `_bor_id()` into a method call rather than a standalone procedure?

**Comment by vasslitvinov** [comment]
Look, that works! Just need to add a path for the reduction code to invoke that.
